### DISCUSSION TITLE
Custom Zims with no date do not Show in Menu

### DISCUSSION THREAD

**Discussion by Axehaft** [iiab/iiab#3830] [Q&A]
I've created zims of some websites I want offline.
I put those zims on a USB and moved them into my IIAB box
They show in the 'Installed Content' list under 'Installed ZIM Files'
But they do not show in my Menu

I can't figure out how to make them show or add them to the home Menu. 

What am I missing??

**Comment by holta** [comment]
Hi @Axehaft !

Can you please run `sudo iiab-diagnostics` and reply here with the URL it generates?

**Comment by Axehaft** [comment]
https://paste.centos.org/view/e544e7ae

**Comment by tim-moody** [comment]
>I can't figure out how to make them show or add them to the home Menu.

To appear on a menu a ZIM needs a corresponding menu def file.

ZIMS downloaded from Kiwix either already have one or one gets generated during the download process. For a ZIM you created you need to make the menu def file manually.

In Admin Console go to Content Menus and select Edit Menu Items. Find a menu item that is similar to the one you created and click the Clone button next to it. This will create a new menu def that you can edit to suit your ZIM. (Make sure you save it.)

Then go to Edit Content Menus and drag your new menu def onto the menu as save.

**Comment by Axehaft** [comment]
I was able to create a menu item this way, but I cannot get it to link to the actual file.  This process is not intuitive. I've included screenshots ... maybe somebody can see what I'm doing wrong.  I did run iiab-make-kiwix-lib ... 
![image_2024-10-14_14-46-36](https://github.com/user-attachments/assets/d83a57ef-b661-4dfc-98f3-2c3a474c2d1b)
![image_2024-10-14_14-47-40](https://github.com/user-attachments/assets/9f95480e-ad4d-4b1d-87bb-b9fc5bdd7135)
![image_2024-10-14_14-47-02](https://github.com/user-attachments/assets/991ca730-fcfa-4eec-8532-69abfab4432f)

**Comment by tim-moody** [comment]
OK. Let's back up. I'll assume your ZIMs are OK, though building them is not trivial.
Did you put them in /library/zims/content/?
Did you run iiab-make-kiwix-lib?
I think this step will have a problem as you have not followed Kiwix's naming convention for ZIMs.

If we look at English Wikipedia as a model you might have a ZIM named wikipedia_en_all_maxi_2024-01.zim
Note the year and month on the end.
In menu definitions we call this wikipedia_en_all_maxi because we don't want to have to change the menu def each time there is a new version.  (it is possible that your underscore will get parsed properly as separating off the version part, but that is certainly not the intent.)

iiab-make-kiwix-lib does two things, it creates a map between the generic name and the specific file name of the ZIM (/library/www/html/common/assets/zim_version_idx.json), and it adds the ZIM to /library/zims/library.xml so kiwix knows to include it. There is also some metadata stored in the ZIM that is expected, such as language code and the ZIM name. The undefined ZIM message means the computed ZIM name was not found.

There is some more info in the Help https://github.com/iiab/iiab-admin-console/blob/master/roles/console/files/help/ContentMenus.rst. But we have not automated the creation of custom ZIMs.

**Comment by Axehaft** [comment]
- I used Zimit to create the zim
- I added them from a USB by moving to iiab.  I assume that adds them to the correct folder?  They show as installed in the Manage Content page.
- I did run iiab-make-kiwix-lib from command line.

It must have something to do with the naming, but it's not clear what the issue is.  Zimit does not add date codes, but I stripped out the end anyway along with the .zim, but that didn't do anything either.  

I can't be the 1st person adding custom zims ... Anyway, I guess I'll just stick to the official ones.

**Comment by tim-moody** [comment]
>I can't be the 1st person adding custom zims ... Anyway, I guess I'll just stick to the official ones.

There are several ways to create zims, but not many try any of them. I have not tried zimit recently, but I would be curious to have a look at one of your zims to see what problems I encounter.

**Comment by Axehaft** [comment]
Sure. I can't upload it here ... it's over a gb.  Suggest where to send it without using personal information?  And Thank you! for taking a look and trying to sort it out!

EDIT:  I have a nextcloud link, just need a method to DM it.

**Comment by tim-moody** [comment]
You can probably find me on LinkedIn.

**Comment by Axehaft** [comment]
Found you.  I don't have Premium Linkedin so I can't DM you there, but I made a connection and my email address is in my profile.

**Comment by tim-moody** [comment]
Got the zim.

* systemctl stop kiwix-serve
* /opt/iiab/kiwix/bin/kiwix-serve -p 3000 /library/zims/content/restoringmayberry.blogspot.com_52da4135.zim start standalone kiwix
* http://iiab-ref:3000/viewer#restoringmayberry.blogspot.com_52da4135/restoringmayberry.blogspot.com/%3Fm%3D1 shows the zim
* restarted kiwix daemon without --nolibrarybutton
* this zim shows on top level page
* http://iiab-ref:3000/kiwix/viewer#restoringmayberry.blogspot.com_52da4135/restoringmayberry.blogspot.com/%3Fm%3D1 accesses zim

So the zim works

* iiab-make-kiwix-lib created what looks like a correct entry in zim_version_idx.json adds entry
* (and added the menu item to the home page)

```
"restoringmayberry.blogspot.com_52da413": {
    "file_name": "restoringmayberry.blogspot.com_52da4135",
    "menu_item": null,
    "article_count": "856",
    "media_count": "1110",
    "size": "1.66G",
    "tags": "_ftindex:yes;_category:other;_pictures:yes;_videos:yes;_details:yes",
    "language": "en",
    "zim_date": "2024-10-11"
  },

```

menu_item is null because none existed when the command was run

OK. Figured it out.

Note that the name of the zim is restoringmayberry.blogspot.com_52da4135.zim.
zim_version_idx.json has an entry for  restoringmayberry.blogspot.com_52da413. So iiab-make-kiwix-lib is removing the final character when it tries to compute the generic name. I guess because it never found the date with the signature hyphen.

My menu def has "zim_name": "restoringmayberry.blogspot.com_52da4135",
If I take off the final '5' it works, because now the lookup in zim_version_idx.json succeeds so we don't get the undefined zim message.

**Comment by tim-moody** [comment]
I guess I'd call this a design flaw and need to handle this case so that menuing works, but we are giving up versions for this class of zims, so I'd prefer that we follow the naming convention even for custom zims. We need a howto that covers the problem.

**Comment by tim-moody** [comment]
Actually, I think this is a bug, but not sure how to fix without breaking something else.

**Comment by tim-moody** [comment]
https://github.com/iiab/iiab/blob/ec348fcc022934baa8010e08c3f6209616300eff/roles/pylibs/templates/iiab_lib.py#L52 and following

assumes there are more than one underscore if no hyphen, based on some old non-canonical zims.

**Comment by tim-moody** [comment]
After some thought, I think the best strategy is to check for the YYYY-MM signature and if found assume there are versions. If not assume no versions and perma ref = file name. This will be less confusing for someone who makes a custom zim, just put the name of the zim in the menu def.
